### PR TITLE
Fix Expr::eval overwrite partial result vector when called with empty selectivity vector

### PR DESCRIPTION
**Pull request by kagamiori**
Summary:
Fuzzer found a bug where SwitchExpr evaluates the second then clause with an empty SelectivityVector because all remaining rows threw during the evaluation of the second condition. However, Expr::eval overwrote the partial result vector with a constant NULL vector and caused incorrect result. This diff fixes the bug by skipping evaluation of then clause in SwitchExpr when there is no selected rows, and make Expr::eval and Expr::evalSimplified only assign a constant NULL vector to result if result is nullptr.

This diff fixes https://github.com/facebookincubator/velox/issues/4585.

Differential Revision: D44920187

### CODE REVIEW

File: velox/expression/Expr.cpp [hunk 1]
```diff
@@ -572,8 +572,10 @@ void Expr::eval(
        topLevel ? (void*)&exprExceptionContext : this});
 
   if (!rows.hasSelections()) {
-    // empty input, return an empty vector of the right type
-    result = BaseVector::createNullConstant(type(), 0, context.pool());
+    if (!result) {
```
Comment (mbasmanova) [inline velox/expression/Expr.cpp:575]: I wonder if it would be better to throw if 'rows' is empty.

```
VELOX_CHECK(rows.hasSelections());
```

Comment (bikramSingh91) [reply velox/expression/Expr.cpp:575]: should we also add similar handling in evalFlatNoNulls() at L563?

File: velox/expression/tests/TryExprTest.cpp [hunk 2]
```diff
@@ -396,5 +401,15 @@ TEST_F(TryExprTest, branchingSpecialForm) {
       "try(switch(always_throws(c6), is_finite(c1), is_finite(c7)))", data);
   expected = makeNullableFlatVector<bool>({true, true, std::nullopt});
   assertEqualVectors(expected, result);
+
+  // Test Switch where the first case produces a partial result while the second
+  // case condition throws at all remaining rows. SwitchExpr may still evaluate
```
Comment (bikramSingh91) [inline velox/expression/tests/TryExprTest.cpp:406]: are we planning to add a change to switch that avoid running the "then" clause for this situation?

Comment (kagamiori) [reply velox/expression/tests/TryExprTest.cpp:406]: Added a check to skip the then clause when there is no remaining rows. Thanks!

File: velox/expression/Expr.cpp (hunk 1 above)
Comment (kagamiori) [reply velox/expression/Expr.cpp:583]: @mbasmanova I tried this but realized that if we simply check rows.hasSelections() in Expr::eval, then we need to have `if (!rows.hasSelection()) { if (!result) { result = BaseVector::createNullConstant(...);} return; }` at all callsites of Expr::eval. In this case, I'd prefer we centralize this handling in Expr::eval only. What do you think?

**Review by bikramSingh91 — APPROVED** [review]
Looks good, thanks for the fix. before merging the PR please reach out to Masha in case she has any concerns regarding this approach.

File: velox/expression/Expr.cpp (hunk 1 above)
Comment (kagamiori) [reply velox/expression/Expr.cpp]: > I wonder if it would be better to throw if 'rows' is empty.
> 
> ```
> VELOX_CHECK(rows.hasSelections());
> ```

Hi @mbasmanova, as we discussed offline, ideally we'd like to make an expression evaluation terminate early if there is no selected rows and check that Expr::eval never gets empty selectivity vector. To achieve this, we need to ensure two things:
1. ExprSet::eval never pass empty selectivity vector to Expr::eval (same for ExprSetSimplified).
2. Any place where selected rows can be reduced during evaluation needs a check that remaining rows is not empty. These places are in SwitchExpr, CoalesceExpr, ConjunctExpr, Expr::evalArgsDefaultNulls(), and Expr::evalArgsWithNulls().

I've checked that these places except SwitchExpr all have proper checks for selectivity vector being non-empty before calling Expr:eval, and this PR adds the check in SwitchExpr. 

However, since this PR adds additional checks that could throw at runtime, I wonder how we should test it to ensure it doesn't affect users. For example, I noticed that the existing unit test ExprTest.testEmptyVectors fails after this change.

One possible way for testing is to add an `always_throw(bool) -> bool` function to expression fuzzer so that some corner cases can be triggered when it's used as a switch condition evaluated in try(). Another thing we can do is to run e2e tests for internal customers. What do you think?

Comment (kagamiori) [reply velox/expression/Expr.cpp]: Hi @mbasmanova, I've tried adding a check that makes Expr::eval and Expr::evalSimplified throw when they receive an empty selectivity vector, and it passed all existing unit tests. However, after discussing with @laithsakka, we think it would be difficult to prove that there won't be such a corner case where Expr::eval is called with an empty selectivity vector during an evaluation. Developers would also have to remember this restriction when they make changes to the evaluator. So I'm changing it back to let Expr::eval and Expr::evalSimplified accept empty selectivity vector to avoid introducing potential bugs.
